bitcoin/examples/sighash.rs

```diff
@@ -0,0 +1,234 @@
+    //Out 0 scriptPubKey.type “multisig” has 3 uncompressed pubkeys
+    let reftx_script_pubkey_bytes = hex!("524104d81fd577272bbe73308c93009eec5dc9fc319fc1ee2e7066e17220a5d47a18314578be2faea34b9f1f8ca078f8621acd4bc22897b03daa422b9bf56646b342a24104ec3afff0b2b66e8152e9018fe3be3fc92b30bf886b3487a525997d00fd9da2d012dce5d5275854adc3106572a5d1e12d4211b228429f5a7b2f7ba92eb0475bb14104b49b496684b02855bc32f5daefa2e2e406db4418f3b86bca5195600951c7d918cdbe5e6d3736ec2abf2dd7610995c3086976b2c0c7b4e459d10b34a316d5a5e753ae");
+    let inp_idx = 0;
+
+    compute_sighash_legacy(raw_tx.as_slice(), inp_idx, Some(reftx_script_pubkey_bytes.as_slice()));
+}
+
+/// Computes segwit sighash for a transaction input that spends a p2wksh output with "witness_v0_keyhash" scriptPubKey.type
+///
+/// # Arguments
+///
+/// * `raw_tx` - spending tx hex
+/// * `inp_idx` - spending tx input index
+/// * `value` - ref tx output value in sats
+fn compute_sighash_p2wpkh(mut raw_tx: &[u8], inp_idx: usize, value: u64) {
+    let tx: bitcoin::Transaction =
+        bitcoin::consensus::Decodable::consensus_decode(&mut raw_tx).unwrap();
+    let inp = &tx.input[inp_idx];
+    let witness = &inp.witness;
+    println!("Witness: {:?}", witness);
+
+    // BIP-141: The witness must consist of exactly 2 items (≤ 520 bytes each). The first one a
+    // signature, and the second one a public key.
+    assert_eq!(witness.len(), 2);
+    let sig_bytes = witness.nth(0).unwrap();
+    let pk_bytes = witness.nth(1).unwrap();
+
+    let sig = bitcoin::ecdsa::Signature::from_slice(sig_bytes).expect("failed to parse sig");
+
+    //BIP-143: "The item 5 : For P2WPKH witness program, the scriptCode is 0x1976a914{20-byte-pubkey-hash}88ac"
+    //this is nothing but a standard P2PKH script OP_DUP OP_HASH160 <pubKeyHash> OP_EQUALVERIFY OP_CHECKSIG:
+    let pk = bitcoin::PublicKey::from_slice(pk_bytes).expect("failed to parse pubkey");
+    let wpkh = pk.wpubkey_hash().expect("compressed key");
+    println!("Script pubkey hash: {:x}", wpkh);
+    let spk = bitcoin::ScriptBuf::new_v0_p2wpkh(&wpkh);
+    let script_code = spk.p2wpkh_script_code().expect("failed to get script code");
+
+    let mut cache = bitcoin::sighash::SighashCache::new(&tx);
+    let sighash = cache
+        .segwit_signature_hash(inp_idx, &script_code, value, sig.hash_ty)
+        .expect("failed to compute sighash");
+    println!("Segwit p2wpkh sighash: {:x}\n", sighash);
+}
+
+/// Computes sighash for a legacy multisig transaction input that spends either a p2sh or a p2ms output.
+///
+/// # Arguments
+///
+/// * `raw_tx` - spending tx hex
+/// * `inp_idx` - spending tx input index
+/// * `script_pubkey_bytes_opt` - Option with scriptPubKey bytes. If None, it's p2sh case, i.e., reftx output's scriptPubKey.type is "scripthash". If Some, it's p2ms case, i.e., reftx output's scriptPubKey.type is "multisig".
+fn compute_sighash_legacy(
```

Comment:
Either I'm not understanding things fully or there is too much going on in this function (see comment below).